### PR TITLE
9.0 purchase order status fields

### PR DESCRIPTION
**Pull request by houssine78**
https://gestion.coopiteasy.be/web#id=3228&view_type=form&model=project.task

### CODE REVIEW

File: partner_contact_address/__openerp__.py [hunk 1]
```diff
@@ -1,37 +1,20 @@
...
-#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-#    GNU Affero General Public License for more details.
-#
-#    You should have received a copy of the GNU Affero General Public License
-#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
-#
-##############################################################################
+# © 2019 Houssine BAKKALI, Coop IT Easy SCRLfs
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
```
Comment (robinkeunen) [inline partner_contact_address/__openerp__.py:5]: We agreed on:
```
# Copyright 2019 Coop IT Easy SCRLfs
# 	Houssine Bakkali <houssine@coopiteasy.be>
# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
```

File: partner_contact_address/models/partner.py [hunk 2]
```diff
@@ -8,7 +8,7 @@
 class ResPartner(models.Model):
```
Comment (robinkeunen) [inline partner_contact_address/models/partner.py:8]: I think it got mixed up with another PR.

File: purchase_order_status/__openerp__.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
+# -*- coding: utf-8 -*-
```
Comment (robinkeunen) [inline purchase_order_status/__openerp__.py:1]: We agreed on:
```
# Copyright 2019 Coop IT Easy SCRLfs
# 	Houssine Bakkali <houssine@coopiteasy.be>
# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
```

File: purchase_order_status/models/purchase.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+
+class PurchaseOrder(models.Model):
+    _inherit = "purchase.order"
+
+    @api.depends('state', 'order_line.qty_invoiced', 'order_line.product_qty')
+    def _get_invoiced(self):
+        precision = self.env['decimal.precision'].precision_get('Product Unit of Measure') #noqa
+        for po in self:
+            invoice_status = 'no'
```
Comment (robinkeunen) [inline purchase_order_status/models/purchase.py]: No need to declare the variable here.

File: purchase_order_status/models/purchase.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+            if po.state in ['purchase', 'done']:
+                if any(float_compare(line.qty_invoiced, line.product_qty, precision_digits=precision) == -1 for line in po.order_line): #noqa
+                    invoice_status = 'to invoice'
+                elif all(float_compare(line.qty_invoiced, line.product_qty, precision_digits=precision) >= 0 for line in po.order_line): #noqa
+                    invoice_status = 'invoiced'
+                else:
+                    invoice_status = 'no'
```
Comment (robinkeunen) [inline purchase_order_status/models/purchase.py]: Nice use of a `any` and `all` but don't abuse the `# noqa` tag, this is more readable with 

```python
if any(
    float_compare(line.qty_invoiced, line.product_qty, precision_digits=precision)
    for line in po.order_line)): 
```
or 
```python
def comp(a, b):
    return float_compare(a, b, precision_digits=precision)
if any(comp(line.qty_invoiced, line.product_qty) == -1 for line in po.order_line): 
    ...
elif all(comp(line.qty_invoiced, line.product_qty) >= 0 for line in po.order_line):
   ...
else:
   ...
```

File: purchase_order_status/models/purchase.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+from openerp import api, fields, models
+from openerp.tools.float_utils import float_compare
+
+
+class PurchaseOrder(models.Model):
+    _inherit = "purchase.order"
+
+    @api.depends('state', 'order_line.qty_invoiced', 'order_line.product_qty')
+    def _get_invoiced(self):
+        precision = self.env['decimal.precision'].precision_get('Product Unit of Measure') #noqa
```
Comment (robinkeunen) [inline purchase_order_status/models/purchase.py]: I know `# noqa` is tempting but ... ;-)

File: purchase_order_status/models/purchase.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+                    invoice_status = 'invoiced'
+                else:
+                    invoice_status = 'no'
+            if invoice_status == 'invoiced':
+                invoices = po.invoice_ids.filtered(
+                            lambda inv: inv.state in ['open', 'paid', 'draft'])
+                paid_invoices = po.invoice_ids.filtered(
+                            lambda inv: inv.state == 'paid')
+                if len(paid_invoices) == len(invoices):
+                    invoice_status = 'paid'
```
Comment (robinkeunen) [inline purchase_order_status/models/purchase.py]: 👍

File: purchase_order_status/models/purchase.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+                delivery_status = None
+                for picking in purchase_order.picking_ids.sorted(
+                        key=lambda r: r.name):
+                    if picking.state == 'done':
+                        if delivery_status != 'to_deliver':
+                            if picking.date_done > done_date:
+                                done_date = picking.date_done
+                                delivery_status = 'delivered'
+                    elif picking.state == 'cancel':
+                        if not delivery_status:
```
Comment (robinkeunen) [inline purchase_order_status/models/purchase.py]: Why not using `and`s? and `and not`s? Then you won't have to declare variables outside the blocks.

File: purchase_order_status/models/purchase.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+                purchase_order.delivery_status = delivery_status
+
+    def _search_delivery_status(self, operator, value):
+        filter_function = {
+            '=': lambda po: po.delivery_status == value,
+            '!=': lambda po: po.delivery_status != value,
+        }
+        purchase_orders = self.search([]).filtered(
+            filter_function[operator]
+        )
```
Comment (robinkeunen) [inline purchase_order_status/models/purchase.py]: waaaaaaaaaaaa

**Review by robinkeunen — CHANGES_REQUESTED** [review]
Ya du très bon stuff! Et 2-3 remarques pour aller plus loin. 👍

File: purchase_order_status/models/purchase.py (hunk 5 above)
Comment (houssine78) [reply purchase_order_status/models/purchase.py]: @robinkeunen readability is relative ;)

Comment (houssine78) [reply purchase_order_status/models/purchase.py]: done

File: purchase_order_status/models/purchase.py (hunk 9 above)
Comment (houssine78) [reply purchase_order_status/models/purchase.py]: t'excite pas j'ai du pomper ça ailleurs ;)

File: purchase_order_status/models/purchase.py (hunk 8 above)
Comment (houssine78) [reply purchase_order_status/models/purchase.py]: this is because we can have several picking with different status for a reception order.

Comment (houssine78) [reply purchase_order_status/models/purchase.py]: anyway I joined two if with an and

**Pull request merged**